content/blog/learnings-from-clean-coder.md

```diff
@@ -0,0 +1,70 @@
+This article can be thought of as a list of learnings from the book Clean Code. By no means this is an exhaustive list, but rather a list of all the lessons which had the most impact while reading the book.
+
+### 1. The Meaning of Clean Code
+
+Uncle Bob starts off the book Clean Code by exploring the term itself. Without stating a single clear-cut definition for the term, the author presents varied definitions from the perspective of different computer scientists. Reading through them, one can figure out common themes and characteristics appear again and again – readability, straightforward, maintainable, reads like a prose, and doesn’t contain duplication. Thus, we infer that clean code has more to do with writing understandable code which is both easily maintainable and changeable, against clever logic which is harder to maintain.
+
+### 2. Duplication is the Evil
+
+Uncle Bob goes on a length about various characteristics of bad code. Amongst them, code duplication holds a special spot, because majority of code problems have their origins linked to duplication. It has inspired many coding patterns, techniques and ideologies. For e.g., the database normalization techniques (1NF, 2NF, 3NF…) by Codd is ultimately about removing redundancy in the database design. In fact, paradigms like OOP, AOP, COP, Structured Programming, are all different ways to deal with duplication. So, one of the primary tasks while writing a piece of software is to be mindful about code duplication.
+
+### 3. One and Only One
+
+This phrase appears quite often in the book Clean Code –
+
+>*“Classes should have only one reason to change”*
+
+>*“A single source file should contain only one language”*
+
+>*“Functions should descend only one abstraction level”*
+
+>*“One assert statement per test”*
+
+In a sense, we can conclude that doing more than one thing at a time, or trying to “juggle” multiple things will ultimately lead to poor code quality and make the code less flexible. By making sure that we keep things simple, it makes the life easier for the entire team.
+
+### 4. F.I.R.S.T Principle for Tests
+
+Writing tests is often seen as something which takes a lot of time and doesn’t serve a real purpose.
+
+>*“I wrote the code; I know it works… Don’t you trust me?”*
+
+However, that’s far from the truth. Writing tests unravels many of the aspects we would have otherwise not uncovered. A good test suite could be a challenge to write, but the F.I.R.S.T. principle can help us out:
+
+* Fast – Tests should run fast. If they are slow, nobody would want to run them.
+* Independent – Each test should be independent of other test, and they should be isolated enough to be run individually.
+* Repeatable – They should be environment independent. Should run on dev, prod, QA, etc.
+* Self-Validating – They test should either pass or fail. Anything in between means more time would be wasted in reading logs and stuff.
```

Comment:
Can you find a way to say the same thing without using the word 'stuff'? Another general rule, avoid ambiguity.